internal/pkg/agent/application/coordinator/coordinator_test.go

```diff
@@ -97,6 +99,180 @@ func waitForState(
+			new: []component.Component{
+				{
+					ID:         "component-three",
+					OutputType: "elasticsearch",
+				},
+			},
+			logtest: func(t *testing.T, logs UpdateStats) {
+				require.Equal(t, 2, len(logs.Components.Removed))
+				require.Equal(t, []string{"component-three"}, logs.Components.Added)
+				require.Equal(t, []string{"kafka"}, logs.Outputs.Removed)
+			},
+		},
+		{
+			name: "test-updated-component",
+			old: []component.Component{
+				{
+					ID:         "component-one",
+					OutputType: "elasticsearch",
+					Units: []component.Unit{
+						{ID: "unit-one"},
+						{ID: "unit-two"},
+						{ID: "unit-x"},
+					},
+				},
+			},
+			new: []component.Component{
+				{
+					ID:         "component-one",
+					OutputType: "elasticsearch",
+					Units: []component.Unit{
+						{ID: "unit-one"},
+						{ID: "unit-two"},
+						{ID: "unit-three"},
+					},
+				},
+			},
+			logtest: func(t *testing.T, logs UpdateStats) {
+				require.Contains(t, logs.Components.Updated[0], "unit-three: added")
+				require.Contains(t, logs.Components.Updated[0], "unit-x: removed")
+			},
+		},
+		{
+			name: "just-change-output",
+			old: []component.Component{
+				{
+					ID:         "component-one",
+					OutputType: "elasticsearch",
+				},
+			},
+			new: []component.Component{
+				{
+					ID:         "component-one",
+					OutputType: "logstash",
+				},
+			},
+			logtest: func(t *testing.T, logs UpdateStats) {
+				require.Equal(t, []string{"elasticsearch"}, logs.Outputs.Removed)
+				require.Equal(t, []string{"logstash"}, logs.Outputs.Added)
+			},
+		},
+		{
+			name: "config-update",
+			old: []component.Component{
+				{
+					ID:         "component-one",
+					OutputType: "elasticsearch",
+					Units: []component.Unit{
+						{
+							ID:     "unit-one",
+							Config: &proto.UnitExpectedConfig{Source: mustNewStruct(t, map[string]interface{}{"example": "value"})},
+						},
+					},
+				},
+			},
+			new: []component.Component{
+				{
+					ID:         "component-one",
+					OutputType: "elasticsearch",
+					Units: []component.Unit{
+						{
+							ID:     "unit-one",
+							Config: &proto.UnitExpectedConfig{Source: mustNewStruct(t, map[string]interface{}{"example": "two"})},
+						},
+					},
+				},
+			},
+			logtest: func(t *testing.T, logs UpdateStats) {
+				require.NotEmpty(t, logs.Components.Updated)
+			},
+		},
+	}
+
+	for _, testcase := range cases {
+
+		t.Run(testcase.name, func(t *testing.T) {
+			testCoord := Coordinator{
+				logger:         logp.L(),
+				componentModel: testcase.new,
+			}
+			testCoord.checkAndLogUpdate(testcase.old)
+
+			obsLogs := logp.ObserverLogs().TakeAll()
+			last := obsLogs[len(obsLogs)-1]
+
+			// extract the structured data from the log message
+			testcase.logtest(t, last.Context[0].Interface.(UpdateStats))
+		})
+
+	}
+
+	logp.ObserverLogs()
+
+	err = logp.Sync()
+	require.NoError(t, err)
+	err = logp.Configure(logp.Config{
+		ToStderr: true,
+	})
+	require.NoError(t, err)
+
```

Comment:
Why are you configuring the logger after the tests?

Reply:
Ah, should add a comment there. There's a bug in `logp` for windows where if refuses to close any open file handles, causing the test to fail when go cleans up the logging directory.

Reply:
Oh wait, that's from before I removed the file logging in the test, I can just get rid of that.